### DISCUSSION TITLE
Is it time to bring people in?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#16] [General]
Hey @hx2A there are pretty big Processing & Python & creative coding communities on Twitter, do you think I could start advertising `py5`?

It is unfortunate that Jeremy Douglass, a huge Processing Python fan, is not very active lately, but at some point we could have a dedicated discourse.processing.org category like p5py has... (Jeremy is a moderator there).

**Comment by hx2A** [comment]
I don't think py5 is ready for that kind of exposure, and I'm not sure I'm ready for that either! Posting something in that processing.py Issue was a big step for me.

My goal for the month of June is to improve the tutorials and how-tos so that basic questions are answered and people can easily get up to speed. I think by that point I'll be more comfortable with advertising the project. Also, using community tools like discourse is important, and I will talk to some people I know in the processing foundation about that.

**Reply by villares** [reply]
I understand and respect your decision.

But also, I don't think you will ever be able to answer the most relevant basic questions if you don't let *other people* try it. It is the curse of knowledge, you will not see the gaps yourself. And I think you should have people helping you with documentation as soon as possible, otherwise you might burnout.

**Reply by hx2A** [reply]
Wait, I meant to say by the end of June I'd like to start advertising it to a wider audience (twitter, etc). Before then I will finish getting the documentation to a better state so that people can easily get up to speed. Currently there's a lot that is confusing or left out. Between the end of June and now, I'd like to get feedback from a smaller group of people so I can iron out the obvious problems.

Next week I will be a part of [ITP Camp](https://tisch.nyu.edu/itp/events/spring-2021/itp-camp-2021) and will share it with that group and see how it goes with them.

> And I think you should have people helping you with documentation as soon as possible, otherwise you might burnout.

Yes, absolutely, this is critical. There's no way I can document everything so that people who are new to Processing and/or Python can use this. How do I get people to help me? Is this by any chance something you'd be willing to help with?

**Reply by villares** [reply]
I'd love to help, and I'm showing this carefully, privately, to some people. Asking if they know people who could help with the MacOS issues. I'm sure I can help a bit, and tap my network of Processing and Python people to find more contributors, this project is awesome. But I have to tell you I'm in very poor shape to contribute directly, lots of teaching obligations, stuck in my PhD research, and really low concentration powers and mental energy as my country sinks into half a million Covid-19 deaths.

**Reply by hx2A** [reply]
> I'd love to help, and I'm showing this carefully, privately, to some people. Asking if they know people who could help with the MacOS issues. I'm sure I can help a bit, and tap my network of Processing and Python people to find more contributors, this project is awesome.

Thank you, I really appreciate that. It means a lot.

I also should mention that outside of github I have no public social media presence. The thought of advertising something I built on twitter kind of freaks me out. I need some time to figure out how to proceed in a way that I will be comfortable with. I don't want to open a pandora's box.

> But I have to tell you I'm in very poor shape to contribute directly, lots of teaching obligations, stuck in my PhD research, and really low concentration powers and mental energy as my country sinks into half a million Covid-19 deaths.

Yes, you have more important priorities to deal with. I totally understand. And it is terribly upsetting to me how COVID is impacting your country and many others. I keep reminding people that the pandemic isn't over until it is over globally.

It happens this project started as an outlet for my pandemic anxieties. I wrote the first prototype right when we went into lockdown. For me, coding helps me relax.

**Reply by tabreturn** [reply]
I'm keen to get involved with documentation. I do have some questions about how you might present content to beginners -- using Module, Class, or Imported mode? Which IDE do we recommend, and so on? Also, could we embed animated sketches in the documentation to show exactly what animated code examples produce? If so, how? ([pyp5js](https://berinhard.github.io/pyp5js/) might prove useful here.

Like @villares, I'm juggling PhD and teaching commitments, but I'm excited about the project. I'll find some time to contribute.

Hoping ITP camp goes well!

**Reply by hx2A** [reply]
> I'm keen to get involved with documentation.

Excellent! Your contribution would be a welcome addition to the emerging py5 community.

> I do have some questions about how you might present content to beginners

You should know I don't have all the answers on how to properly develop the documentation and content for py5. I'm hoping to get feedback and advice from others on what to do in the short term and long term. I'll start another thread for that.

> -- using Module, Class, or Imported mode?

Definitely not class mode, that is more advanced. Coders need to know about object oriented programming, etc, and that can get tough. @villares had some good ideas on this about reducing the information overload for beginners by starting them out on imported mode, with the soon-to-be-released py5 feature that drops the requirement that calls to `size()` go in an explicit `settings()` function. I think some kind of "py5bot" tool would be good also.

But, the answer might be different for different groups of learners. For some folks who already know python, module mode might be the most natural place to start.

> Which IDE do we recommend, and so on?

Good question. I'd like for py5 to be flexible enough that people can make their own choices around this. I have looked into this and would suggest [Thonny](https://thonny.org/) and [Mu](https://codewith.mu/) might be interesting places to start. Perhaps either can be customized to work well with py5? If you want to stick with Jupyter, have a look at [nteract](https://nteract.io/). That is a Jupyter client that does not get enough attention.

Keep in mind that none of these will do the kind of syntax highlighting you would want for py5 in imported mode. It won't know that functions like `fill` or `rect` are special and should have a different color than other functions. I do know that Jupyter notebook clients use the same syntax highlighting JavaScript library that https://editor.p5js.org/ uses for its syntax hightlighting, so perhaps someone can customize the jupyter client frontend by learning from the p5js web editor.

> Also, could we embed animated sketches in the documentation to show exactly what animated code examples produce? If so, how? (pyp5js might prove useful here.

There are some sketch hooks that save frames from a running sketch. You can use [py5_tools.animated_gif()](http://py5.ixora.io/reference/animated_gif/) to create an animated gif or perhaps [py5_tools.save_frames()](http://py5.ixora.io/reference/save_frames/) or [py5_tools.capture_frames()](http://py5.ixora.io/reference/capture_frames/) to make a movie you upload to youtube.

> Like @villares, I'm juggling PhD and teaching commitments, but I'm excited about the project. I'll find some time to contribute.

> Hoping ITP camp goes well!

Lately I've been consumed by family stuff and haven't done as much with ITP Camp as I would have liked. But, it isn't over yet, and I'm eager to turn things around by this weekend!

**Reply by tabreturn** [reply]
I wrote a [book on Processing.py](https://nostarch.com/learn-python-visually). Many of the [tutorials on the Processing website(s)](https://processing.org/tutorials/) are extracted/adapted from books (they have a short notice where this is applicable). I'm very happy to see my material used like this.

> Definitely not class mode ... soon-to-be-released py5 feature that drops the requirement that calls to size() ...

Agreed -- I'm a fan of beginner tutorials that start with Imported mode, static sketches at that (with no `setup()` and `draw()` functions), then animation is introduced later. Glad to hear you're dropping the requirement for `size()` in the `settings()` block :+1: 

>... would suggest Thonny and Mu might be interesting places to start. Perhaps either can be customized to work well with py5 ...

I'm a big fan of Thonny. I've done some [experimenting combining with Thonny and py5](https://tabreturn.github.io/code/python/thonny/2021/06/21/thonny_and_py5.html), which works great.

> Keep in mind that none of these will do the kind of syntax highlighting you would want for py5 in imported mode ...

I've already hacked Thonny to do syntax highlighting for py5 in Imported mode; code-completion looks quite easy too. But those are quick and dirty approaches. I think the [Thonny plug-in](https://github.com/thonny/thonny/wiki/Plugins) system could be the solution. We might also use the plug-in system to add useful Processing PDE features (a colour mixer, pyde file extensions, etc.)

We might even combine a hacked Thonny, Java runtime, and other py5 dependencies into a [portable app](https://en.wikipedia.org/wiki/Portable_application). Super easy for beginners to just download, click, and start coding; runs off a USB; all that good stuff. Processing is distributed like this.

> ... use py5_tools.animated_gif() to create an animated gif or perhaps py5_tools.save_frames() or py5_tools.capture_frames() ...

With pyp5js we'll get animation and interactivity :sunglasses:. I built a [website that batch transcripts the Processing.py examples](http://pyde.org/) using pyp5js to run in a browser. Just something to consider ...

**Reply by hx2A** [reply]
> I wrote a book on Processing.py. Many of the tutorials on the Processing website(s) are extracted/adapted from books (they have a short notice where this is applicable). I'm very happy to see my material used like this.

Awesome! Your ability to teach Python to beginners far exceeds mine.

I didn't know that the tutorials were adapted from books.

Someday, I'd also like to write a book.

> I'm a big fan of Thonny. I've done some experimenting combining with Thonny and py5, which works great.

Cool, and thanks for the blog post!

There are a few things I need to comment on:

* A few hours ago I moved the py5 blog posts to be with my regular blog, so that blog link now goes to https://ixora.io/tags/py5/. Having a separate blog seemed messy and keeping 2 of them up to date is inefficient. I was hoping to make that switch before anyone linked to it...guess I was 1 day too slow. :) That new link will only show posts with the py5 tag on my main blog.
* There probably won't be a compatability mode like you describe where py5 executes Python code with functions in camel case. But I am working on some translation tools that attempt to convert processing.py or camel case Python code to use function names that py5 expects. I ran it on some of @villares 's processing.py code and about 80% of it worked perfectly, without any additional manual changes. Note there are more differences between processing.py and py5 than the camel case vs snake case issue, so getting py5 to run processing.py code as-is would be impossible.
* I love how easy it was to get imported mode to run in thonny. I had no idea when I made that `run_sketch` utility that it would be that easy to integrate it into the thonny editor. I'm very excited about this.
* I've already merged the code for dropping the `settings()` requirement so it will definitely be in the next release.
* A poorly documented feature is the ability to [customize error messages](http://py5.ixora.io/tutorials/exceptions-debugging-performance/#Customizing-Error-Messages). I believe Mu has a feature where it provides advice when you get an exception (or was it Thonny?). Anyhow, all of the error messages can be customized to provide beginner friendly advice. 

> I've already hacked Thonny to do syntax highlighting for py5 in Imported mode; code-completion looks quite easy too. But those are quick and dirty approaches. I think the Thonny plug-in system could be the solution. We might also use the plug-in system to add useful Processing PDE features (a colour mixer, pyde file extensions, etc.)

> We might even combine a hacked Thonny, Java runtime, and other py5 dependencies into a portable app. Super easy for beginners to just download, click, and start coding; runs off a USB; all that good stuff. Processing is distributed like this.

This is all amazing, and an excellent path forward for making py5 accessible to beginners! Thank you so much for doing this. I think getting Thonny + py5 to a good state will be much easier than trying to get it working in the Processing IDE. Wouldn't it be hard to get the Processing IDE to manage all of py5's Python package dependencies?

**Reply by tabreturn** [reply]
Thanks for the heads-up -- I've made the necessary changes to my blog post.

The customisable error messages are neat touch :+1:   

> Wouldn't it be hard to get the Processing IDE to manage all of py5's Python package dependencies?

I suspect that Thonny + p5 is the easiest way to ship a py5 IDE. TBH, I'm not sure how difficult it'll be integrating py5 into the Processing IDE. At the very least, I think that py5's approach offers a feasible alternative to Processing.py's Jython implementation.

**Comment by sawula** [comment]
Hi,  

HS teacher.  I've been emailing hx2a since mid June...didn't realize that the party was over here!  

About 10 yrs ago I went to ITP Camp, Eyeo, and was a minor part of (Ira Greenberg and Dianna Xu)'s NSF grant about teaching CS1 using creative coding.  I used processing.py with a data viz interest.  More recently teaching a data science elective using Jupyter and this year, JupyterHub with support from 2i2c.  So I've got a hub now that runs py5 and I'm going to trial it with a tiny number (6) of 8th graders shortly.  Thinking through how to blend the creative coding approach with introductory data science topics (think of UC Berkeley's Data 8 as a reference www.inferentialthinking.com).

The Berkeley data science education workshop begat a Slack channel.  2i2c uses one, too. I would welcome that.

**Reply by hx2A** [reply]
Welcome @sawula ! We have an emerging py5 community here.

> About 10 yrs ago I went to ITP Camp

I've been at ITP Camp every year since 2014. We might know some of the same people.

> So I've got a hub now that runs py5 and I'm going to trial it with a tiny number (6) of 8th graders shortly.

We're very eager to hear how this goes! The other folks posting here also have interests in the educational role of py5.

> The Berkeley data science education workshop begat a Slack channel. 2i2c uses one, too. I would welcome that.

I was thinking of starting a Discord server for py5, which is comparable to a Slack channel. My understanding is that Slack is more popular with businesses but Discord is more popular with open source projects. If I created a Discord channel, would people be interested in using it?

**Reply by hx2A** [reply]
@sawula how did it go this week with the students?

**Reply by sawula** [reply]
Everything went well.  

- Technically, no problems with py5.  I only asked students to make very simple static sketches (background and a few rectangles), but there were no surprises.
- Pedagogically, the experiment of giving them tasks they could figure out on their own with minimal explanation also worked.  As did using a separate notebook for each task.

What's next?  Still feeling my way around with curriculum.  A fun and productive turning over of ideas.  With a little anxiety and time pressure thrown in.

- I think it would be nice in imported mode to have `sketch_portal()` available from the start
- I had been struggling with getting a slider widget to interact with an animated sketch but I just sorted that out. So cool.  
- I'm a little confused about how the sketch_portal does not clear/reset when being run again.  The outputs just seem to accrue.  Oh, now I see.  It has to do with the mysteries of exit_sketch().  Even if I set a time_limit on the sketch which is met, it stops updating, but doesn't actually exit.

**Reply by hx2A** [reply]
Awesome, I am glad it went well!

> I think it would be nice in imported mode to have sketch_portal() available from the start

Good idea. I just took care of it. Next release due out soon.

> I had been struggling with getting a slider widget to interact with an animated sketch but I just sorted that out. So cool.

The other IPython widgets are a great tool that pair well with py5, especially when run on Jupyter Hub. The Sketch Portal doesn't support mouse or keyboard inputs but the many widgets create new opportunities. You can build little applications to run with Viola on Jupyter Hub. There's a lot of potential there.

> I'm a little confused about how the sketch_portal does not clear/reset when being run again. The outputs just seem to accrue. Oh, now I see. It has to do with the mysteries of exit_sketch(). Even if I set a time_limit on the sketch which is met, it stops updating, but doesn't actually exit.

The Sketch portal is really just a "portal" or view into what is happening elsewhere on the server. It is not the Sketch itself. I agree that exiting is going to be an issue that needs to be addressed, especially if you do something in Viola. There is the tendency to just close the Viola tab to exit but that doesn't shutdown the Sketch or the portal. You can't have multiple Sketches and portals running at the same time on Jupyter Hub without killing performance. This will be confusing for beginners.

I need to do a release (this weekend?) with what I have accomplished over the past month or so. There's a second Jupyter kernel called py5bot and a new magic called %py5bot that is available in the py5 kernel (imported mode). People will be able to put calls to `size` in `setup()` and I made many improvements to error message reporting. And best of all, I made massive improvements with my OSX issues.

After this release I'll see what I can do to improve the Sketch Portal. At a minimum there should be a button associated with it for stopping the sketch.

**Reply by hx2A** [reply]
> After this release I'll see what I can do to improve the Sketch Portal. At a minimum there should be a button associated with it for stopping the sketch.

@sawula this morning I was feeling inspired and added this button. It was just a few lines of code.

**Comment by sawula** [comment]
That's awesome!

Have you seen this?  t.ly/diH5
There are lots of cool things happening in this space now!

**Reply by hx2A** [reply]
I did see that and am very happy about that. I like that they have numpy working in there too. Very impressive.

I thought it would be a far-fetched idea for me to think that one day I'd be able to get py5 working in JupyterLite but I then found several projects that are attempting to get java byte code working in web assembly. The lead developer for jpype is very smart and motivated, so when it is possible to get jpype working JupyterLite, I'm confident he'll make it happen.

**Comment by hx2A** [comment]
@villares @sawula I just did the latest release, which includes py5bot and the ability to combine settings and setup into one function.

My brain is fried from writing documentation all day. I'm going to wait a few days to make sure no bugs emerge, and if it still looks OK, I'll start promoting py5 on twitter and elsewhere.

Everything is working in mybinder with https://github.com/hx2A/py5examples.  Please have a look and provide feedback!

**Reply by hx2A** [reply]
@villares @tabreturn I botched the release. There are some bugs in py5bot that have no business being there at all. :-1: 

You can install the latest development version directly from the py5 github repo with this command:

```
pip install git+https://github.com/hx2A/py5.git@0.5a1.dev0#egg=py5
```
I pushed the latest generated code to the 0.5a1.dev0 branch in that repo, which is why you can install it from there. Moving forward I will continue to do this so people who want to test the very latest version can do so.

I need to develop some unit tests for this project or I am going to get killed by bad releases.

**Comment by sawula** [comment]
Those changes might seem small, but I appreciate them. Makes the initial
encounters with py5 a little more streamlined.  Who recommended them?  I
was wondering about settings() and setup() myself but didn't dare to ask!

**Reply by hx2A** [reply]
You are welcome! Both @villares and @tabreturn recommended the settings / setup feature. The idea is to reduce the cognitive load on beginner coders by reducing the number of things that need to be done or understood. It took a lot of thought to come up with an approach that works correctly!

**Comment by hx2A** [comment]
@villares I finally started using the py5 twitter account, py5coding. I see you already re-tweeted it...thank you!

Up until a few days ago I had no public social media presence outside of github. I have a lot to learn about how to function on twitter. Any suggestions for what I can do better would be helpful.

That twitter account is for py5 and not me personally. I do have a second twitter account that I will start using for that at a later date. My instagram account is now public, and is jim_hx2a, if you are on that platform. Currently very little of it relates to py5 but that will change moving forward.

**Reply by villares** [reply]
Cheers, Jim!

The Japanese Processing community is very friendly, and some of them use Processing Python mode, like [ryotakob](https://twitter.com/ryotakob), I have DM them (sent a direct message, to the private inbox) and they have retweeted your tweet. Feel free to DM me if you feel like it. 

I had tipped some friends about py5 and now I'm doing the second round telling the ones on twitter about the account.

I'm huge Allisson Parrish fan, so I tried to contact her on Mastodon, only for to her reply she was already helping :rofl:

**Reply by hx2A** [reply]
> The Japanese Processing community is very friendly, and some of them use Processing Python mode, like [ryotakob](https://twitter.com/ryotakob), I have DM them (sent a direct message, to the private inbox) and they have retweeted your tweet. Feel free to DM me if you feel like it.

I'm looking forward to meeting them! And I figured out how to use the translate feature in Twitter.

> I'm huge Allisson Parrish fan, so I tried to contact her on Mastodon, only for to her reply she was already helping 🤣

Allison is wonderful, I was in one of her classes at NYU ITP. We've talked over zoom a few times about py5 and she inspired several important features that I use all the time. It was because of her I realized how important it was to get py5 working well in Jupyter.

**Reply by villares** [reply]
FYI: @hx2A, I understand you are careful/worrisome about social media exposure and interaction, but sometimes DMs (inbox, private Direct Messages) can be helpful. On Twitter, by default, only people you follow can DM you.

**Reply by hx2A** [reply]
> FYI: @hx2A, I understand you are careful/worrisome about social media exposure and interaction, but sometimes DMs (inbox, private Direct Messages) can be helpful. On Twitter, by default, only people you follow can DM you.

@villares I assume you are referring to my reply to your tweet. I didn't do anything wrong there, did I? Should I have sent a DM instead?

I didn't know DMs were an option or how that worked. I see there is a message button when I go to your profile page that I don't see on random profile pages. That is how I send you a DM? Thank you for pointing this out. Is this a feature people use a lot? I can see how this would be useful.

I am careful or hesitant about online stuff but this is something I both need and want to learn about and get comfortable with. This is something about myself that I want to change.

**Reply by villares** [reply]
Oh, nothing wrong at all @hx2A!  You will see the message button only on profiles that follow you (like me), that's exactly it. But as you follow very few people we can't DM you. I wanted, for instance, to give you "in private" some context about Tariq who reacted about the np_pixels tweet, he is a wonderful person who self publishes books about Python and is also into Creative Coding, teaching and mathematics.

**Reply by villares** [reply]
Hey @tabreturn, why don't you join us on Twitter, instead of finishing your PhD :rofl: 
(I know, I know, I should try to finish mine instead of trying to derail other people's PhDs)

**Reply by tabreturn** [reply]
Time to try out that Twitter account I signed up for years ago and never got around to using ...

**Reply by hx2A** [reply]
@villares got it, that makes sense. Eventually when I get more comfortable with this I'll start using a second twitter account that will be for me personally, and py5coding will represent the py5 project. I'm starting to see the value of this platform.
